### PR TITLE
Port last remaining local serializer test.

### PR DESCRIPTION
**Pull request by rsgowman**
Also involves porting:
- MutationBatch (partial)
- LocalSerializer::[En|De]codeMutationBatch
- Serializer::[En|De]codeMutation
- Serializer::[En|De]codePrecondition
- Serializer::[En|De]codeDocumentMask

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/model/mutation.h [hunk 1]
```diff
@@ -178,18 +186,28 @@ class Mutation {
       const MaybeDocument* base_doc,
       const Timestamp& local_write_time) const = 0;
 
+  friend bool operator==(const Mutation& lhs, const Mutation& rhs) {
```
Comment (zxu123) [inline Firestore/core/src/firebase/firestore/model/mutation.h]: optional: I feel moving the inline below close to operator!= and keep here just a friend declaration is easier to read. I cannot find any style guide on this and thus optional.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 2]
```diff
@@ -491,6 +501,199 @@ std::unique_ptr<Document> Serializer::DecodeDocument(
...
+      MakeArray<google_firestore_v1_DocumentTransform_FieldTransform>(count);
+        int i = 0;
+        for (const FieldTransform& field_transform :
+      transform.field_transforms()) { result.transform.field_transforms[i] =
+      EncodeFieldTransform(field_transform); i++;
+        }
+        return result;
+      */
+
+    case Mutation::Type::kDelete:
```
Comment (zxu123) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: The style is not consistent. I know the kPatch has new local variable and asks for {. How about make every case use {.

File: Firestore/core/src/firebase/firestore/local/local_serializer.cc [hunk 3]
```diff
@@ -137,11 +156,8 @@ firestore_client_NoDocument LocalSerializer::EncodeNoDocument(
 
 std::unique_ptr<NoDocument> LocalSerializer::DecodeNoDocument(
     Reader* reader, const firestore_client_NoDocument& proto) const {
-  if (!reader->status().ok()) return nullptr;
```
Comment (zxu123) [inline Firestore/core/src/firebase/firestore/local/local_serializer.cc]: This and below are the checks I talked this morning. I guess now the check for status().ok() is centralized somewhere and thus individual deserializer does not check it anymore. Just want to confirm.

File: Firestore/core/test/firebase/firestore/local/local_serializer_test.cc [hunk 4]
```diff
@@ -205,11 +237,106 @@ class LocalSerializerTest : public ::testing::Test {
     return bytes;
   }
 
+  void ExpectSerializationRoundTrip(
+      const MutationBatch& model,
+      const ::firestore::client::WriteBatch& proto) {
+    std::vector<uint8_t> bytes = EncodeMutationBatch(&serializer, model);
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/test/firebase/firestore/local/local_serializer_test.cc]: This is neat. When taken with the analogous functions within the remote serializer, there's a few problems though:
a) encoder() sometimes takes more than one parameter. (eg EncodeDocument(key, value)). I suppose you could do something with std::bind...
b) many of these functions have additional logic, implying you mostly wouldn't be able to get rid of the extra function.
c) Sometimes the input and expected protos are not the same type.

I tried it anyways, but it didn't seem to be a net win; there's just too much that varies, and not enough that stays the same.

[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/test/firebase/firestore/local/local_serializer_test.cc]: Point a) can probably be solved with variadics, but it does sound like it's more trouble than it's worth, all in all.

**Pull request merged**